documentation/IORuntimeInternals.md

```diff
@@ -0,0 +1,341 @@
+language standard in its clauses 12 (I/O statements) and 13 (`FORMAT`).
+It's a complicated collection of language features:
+ * Files can comprise *records* or *streams*.
+ * Records can be fixed-length or variable-length.
+ * Record files can be accessed sequentially or *direct* (random access).
+ * Files can be *formatted*, or *unformatted* raw bits.
+ * `CHARACTER` scalars and arrays can be used as if they were
+fixed-length formatted sequential record files.
+ * Formatted I/O can be under control of a `FORMAT` statement
+or `FMT=` specifier, *list-directed* with default formatting chosen
+by the runtime, or `NAMELIST`, in which a collection of variables
+can be given a name and passed as a group to the runtime library.
+ * Sequential records of a file can be partially processed by one
+or more *non-advancing* I/O statements and eventually completed by
+another.
+ * `FORMAT` strings can manipulate the position in the current
+record arbitrarily, causing re-reading or overwriting.
+ * Floating-point output formatting supports more rounding modes
+than the IEEE standard for floating-point arithmetic.
+
+The Fortran I/O runtime support library is written in C++'17, and
+uses some C++'17 standard library facilities, but it is intended
+to not have any link-time dependences on the C++ runtime support
+library or any LLVM libraries.
+This is important because there are at least two C++ runtime support
+libraries, and we don't want Fortran application builders to have to
+build multiple versions of their codes; neither do we want to require
+them to ship LLVM libraries along with their products.
+
+Consequently, dynamic memory allocation in the Fortran runtime
+uses only C's `malloc()` and `free()` functions, and the few
+C++ standard class templates that we instantiate in the library have been
+modified with optional template arguments that override their
+allocators and deallocators.
+
+Conversions between the many binary floating-point formats supported
+by f18 and their decimal representations are performed with the same
+template library of fast conversion algorithms used to interpret
+floating-point values in Fortran source programs and to emit them
+to module files.
+
+Overview of Classes
+===================
+
+A suite of C++ classes and class templates are composed to construct
+the Fortran I/O runtime support library.
+They (mostly) reside in the C++ namespace `Fortran::runtime::io`.
+They are summarized here in a bottom-up order of dependence.
+
+The header and C++ implementation source file names of these
+classes are in the process of being vigorously rearranged and
+modified; use `grep` or an IDE to discover these classes in
+the source for now.  (Sorry!)
+
+`Terminator`
+----------
+A general facility for the entire library, `Terminator` latches a
+source program statement location in terms of an unowned pointer to
+its source file path name and line number and uses them to construct
```

Comment:
I don't think it's worth having, since it's rare for Fortran programs to have multiple I/O statements on the same line, and I don't want every program to incur the overhead of a column number argument.